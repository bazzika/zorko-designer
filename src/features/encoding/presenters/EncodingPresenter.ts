import {PositionChannelState} from './PositionChannelPresenter';

export enum EncodingChannelNames {
  X = 'x',
  Y = 'y'
}

export interface EncodingState {
  x: PositionChannelState | null;
}

export class EncodingPresenter {
  private state: EncodingState;

  static create(state?: EncodingState) {
    return new EncodingPresenter(state);
  }

  constructor(state?: EncodingState) {
    if (!state) {
      this.reset();
    } else {
      this.state = state;
    }
  }

  editByName(name, modificationCallback) {
    // TODO: rethink modification of presenter
    const channel = this.state[name];
    this.state[name] = modificationCallback(channel);

    return this;
  }

  reset() {
    this.state = {
      x: null
    };
    return this;
  }

  all() {
    // move order to analytic board
    return [];
  }

  toState(): EncodingState {
    return this.state;
  }
}
