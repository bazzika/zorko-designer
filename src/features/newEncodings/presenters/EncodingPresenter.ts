import {BaseStatePresenter} from '../../../core/BaseStatePresenter';
import {EncodingState} from './EncodingState';

export enum EncodingChannelNames {
  X = 'x',
  Y = 'y',
  COLOR = 'color'
}

export type EncodingChannelType =
  | EncodingChannelNames.X
  | EncodingChannelNames.Y
  | EncodingChannelNames.COLOR
  | string;

export class EncodingPresenter extends BaseStatePresenter<EncodingState> {
  static create(state?: EncodingState) {
    return new EncodingPresenter(state);
  }

  getDefaultState(): EncodingState {
    return {};
  }

  setChannel(name: EncodingChannelType, value: string): this {
    this.state[name] = value;
    return this;
  }

  setX(x: string): this {
    this.state.x = x;
    return this;
  }

  setY(y: string): this {
    this.state.y = y;
    return this;
  }

  setColor(color: string): this {
    this.state.color = color;
    return this;
  }

  getX() {
    return this.state.x;
  }

  getY() {
    return this.state.y;
  }

  getColor() {
    return this.state.color;
  }
}
