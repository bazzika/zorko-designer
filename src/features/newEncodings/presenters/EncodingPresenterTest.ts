import {EncodingChannelNames, EncodingPresenter} from './EncodingPresenter';

describe('Encoding Presenter', () => {
  let presenter: EncodingPresenter;

  beforeEach(() => {
    presenter = EncodingPresenter.create();
  });

  it('creates default', () => {
    expect(EncodingPresenter.create().toState()).toMatchSnapshot();
  });

  it('sets x', () => {
    expect(presenter.setX('bar/x').toState()).toMatchSnapshot();
  });

  it('sets y', () => {
    expect(presenter.setY('bar/y').toState()).toMatchSnapshot();
  });

  it('sets color', () => {
    expect(presenter.setColor('bar/color').toState()).toMatchSnapshot();
  });

  it('gets y', () => {
    expect(presenter.setY('bar/y').getY()).toEqual('bar/y');
  });

  it('gets x', () => {
    expect(presenter.setX('bar/x').getX()).toEqual('bar/x');
  });

  it('gets color', () => {
    expect(presenter.setColor('bar/color').getColor()).toEqual('bar/color');
  });

  it('sets channel by name', () => {
    expect(presenter.setChannel(EncodingChannelNames.X, 'bar/x').getX()).toEqual('bar/x');
    expect(presenter.setChannel(EncodingChannelNames.Y, 'bar/y').getY()).toEqual('bar/y');
    expect(presenter.setChannel(EncodingChannelNames.COLOR, 'bar/color').getColor()).toEqual(
      'bar/color'
    );
  });
});
