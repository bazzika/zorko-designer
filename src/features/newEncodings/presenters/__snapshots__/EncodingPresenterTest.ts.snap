// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Encoding Presenter creates default 1`] = `Object {}`;

exports[`Encoding Presenter sets color 1`] = `
Object {
  "color": "bar/color",
}
`;

exports[`Encoding Presenter sets x 1`] = `
Object {
  "x": "bar/x",
}
`;

exports[`Encoding Presenter sets y 1`] = `
Object {
  "y": "bar/y",
}
`;
